File: sketch-core/src/main/java/com/github/panpf/sketch/util/pool/BaseKeyPool.java
package com.github.panpf.sketch.util.pool;

import androidx.annotation.NonNull;

import java.util.ArrayDeque;
import java.util.Queue;

abstract class BaseKeyPool<T extends Poolable> {

    private static final int MAX_SIZE = 20;
    private final Queue<T> keyPool = new ArrayDeque<>(MAX_SIZE);

    @NonNull
    T get() {
        T result = keyPool.poll();
        if (result == null) {
            result = create();
        }
        return result;
    }

    public void offer(@NonNull T key) {
        if (keyPool.size() < MAX_SIZE) {
            keyPool.offer(key);
        }
    }

    @NonNull
    abstract T create();
}

File: sketch-core/src/main/java/com/github/panpf/sketch/util/pool/BaseKeyPool.kt
package com.github.panpf.sketch.util.pool

import java.util.ArrayDeque
import java.util.Queue

internal abstract class BaseKeyPool<T : Poolable?> {

    private val keyPool: Queue<T> = ArrayDeque(MAX_SIZE)

    fun get(): T {
        return keyPool.poll() ?: create()
    }

    fun offer(key: T) {
        if (keyPool.size < MAX_SIZE) {
            keyPool.offer(key)
        }
    }

    abstract fun create(): T

    companion object {
        private const val MAX_SIZE = 20
    }
}
